src/game_strings.cpp

```diff
@@ -0,0 +1,223 @@
+
+	int index;
+	std::string result;
+	std::string str = static_cast<std::string>(Get(params.string_id));
+
+	std::stringstream ss(str);
+
+	while (offset >= 0 && std::getline(ss, result)) { offset--; }
+
+	offset = ss.rdbuf()->in_avail();
+
+	Set(params, static_cast<Str_t>(ss.str().substr(str.length() - offset)));
+	params.string_id = string_out_id;
+	return Set(params, static_cast<Str_t>(result));
+}
+
+Game_Strings::Str_t Game_Strings::ExMatch(Str_Params params, std::string expr, int var_id, int begin, int string_out_id) {
+	int var_result;
+	Str_t str_result;
+	std::smatch match;
+
+	if (params.extract) {
+		expr = static_cast<std::string>(Extract(static_cast<Str_t>(expr), params.hex));
+	}
+
+	std::string base = static_cast<std::string>(Get(params.string_id)).erase(0, begin);
+	std::regex r(expr);
+
+	std::regex_search(base, match, r);
+
+	var_result = match.position() + begin;
+	Main_Data::game_variables->Set(var_id, var_result);
+
+	str_result = static_cast<Str_t>(match.str());
+	if (string_out_id > 0) {
+		params.string_id = string_out_id;
+		Set(params, str_result);
+	}
+	return str_result;
+}
+
+const Game_Strings::Strings_t& Game_Strings::RangeOp(Str_Params params, int string_id_1, Str_t string, int op, int args[]) {
+	if (EP_UNLIKELY(ShouldWarn(params.string_id))) {
+		WarnGet(params.string_id);
+	}
+	if (EP_UNLIKELY(ShouldWarn(string_id_1))) {
+		WarnGet(string_id_1);
+	}
+	if (params.string_id <= 0 && string_id_1 <= 0) { return GetData(); }
+
+	// maniacs just ignores if only one of the params is <= 0
+	if (params.string_id <= 0) { params.string_id = 1; }
+	if (string_id_1 <= 0) {	string_id_1 = 1; }
+
+	// swap so that id_0 is < id_1
+	if (params.string_id > string_id_1) {
+		params.string_id = params.string_id ^ string_id_1;
+		string_id_1 = string_id_1 ^ params.string_id;
+		params.string_id = params.string_id ^ string_id_1;
+	}
+
+	if (EP_UNLIKELY(string_id_1 > static_cast<int>(_strings.size()))) {
+		_strings.resize(string_id_1, "");
+	}
+
+	for (int start = params.string_id; params.string_id <= string_id_1; params.string_id++) {
+		switch (op) {
+		case 0:  Asg(params, string); break;
+		case 1:  Cat(params, string); break;
+		case 2:  ToNum(params, args[0] + (params.string_id - start)); break;
+		case 3:  GetLen(params, args[0] + (params.string_id - start)); break;
+		case 4:  InStr(params, static_cast<std::string>(string), args[1], args[2]); break;
+		case 5:  params.string_id += Split(params, static_cast<std::string>(string), args[1], args[2]); break;
+		case 8:  break; // range case not applicable for popLine; see case in game_interpreter.cpp
+		case 9:  ExMatch(params, static_cast<std::string>(string), args[1] + (params.string_id - start), args[2]); break;
+		case 10: ExMatch(params, static_cast<std::string>(string), args[1] + (params.string_id - start), args[2], args[3]); break;
+		}
+	}
+	return GetData();
+}
+
+Game_Strings::Str_t Game_Strings::PrependMin(Str_t string, int min_size, char c) {
```

Comment:
why is the ``PrependMin`` required?

Reply:
Some string assignment operations have an option for creating a "minimum length" for the saved string. For example, when assigning a string as a variable, such as `t[1] .asg v[1]`, you can also optionally specify a `.min`, like `t[1] .asg v[1] .min 5`, to give the string a minimum length. In the case of a number, it will prepend `0`s, and for a normal string, it will prepend whitespace.

e.g.
```
v[1] = 3
t[1] .asg v[1] .min 5
@msg.show "\t[1]"
```
will result in displaying `00003`.